app/code/Magento/QuoteBundleOptions/Model/Cart/BuyRequest/BundleDataProvider.php

```diff
@@ -0,0 +1,76 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\QuoteBundleOptions\Model\Cart\BuyRequest;
+
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Quote\Model\Cart\BuyRequest\BuyRequestDataProviderInterface;
+use Magento\Quote\Model\Cart\Data\CartItem;
+
+/**
+ * Data provider for bundle product buy requests
+ */
+class BundleDataProvider implements BuyRequestDataProviderInterface
+{
+    private const OPTION_TYPE = 'bundle';
+
+    /**
+     * @inheritdoc
+     *
+     * @throws LocalizedException
+     */
+    public function execute(CartItem $cartItem): array
+    {
+        $bundleOptionsData = [];
+
+        foreach ($cartItem->getSelectedOptions() as $optionData) {
+            // phpcs:ignore Magento2.Functions.DiscouragedFunction
+            $optionData = \explode('/', base64_decode($optionData->getId()));
+
+            if ($this->isProviderApplicable($optionData) === false) {
+                continue;
+            }
+            $this->validateInput($optionData);
+
+            [, $optionId, $optionValueId, $optionQuantity] = $optionData;
+            $bundleOptionsData['bundle_option'][$optionId] = $optionValueId;
+            $bundleOptionsData['bundle_option_qty'][$optionId] = $optionQuantity;
+        }
+
+        return $bundleOptionsData;
+    }
+
+    /**
+     * Checks whether this provider is applicable for the current option
+     *
+     * @param array $optionData
+     * @return bool
+     */
+    private function isProviderApplicable(array $optionData): bool
+    {
+        if ($optionData[0] !== self::OPTION_TYPE) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Validates the provided options structure
+     *
+     * @param array $optionData
+     * @throws LocalizedException
+     */
+    private function validateInput(array $optionData): void
+    {
+        if (count($optionData) !== 4) {
+            $errorMessage = __('Wrong format of the entered option data. ' .
+                'Must be "bundle/option_id/option_value_id/option_quantity"');
```

Comment:
We cannot expose implementation details to Customer. "id" should be interpreted as one unit. Current implementation with encoded value just a workaround until UUID will be introduced in Magento (see details in https://github.com/magento/architecture/blob/master/design-documents/graph-ql/coverage/add-items-to-cart-single-mutation.md#options)

Maybe we can log it just for debug purpose and throw a generic error message aka "Provided option %1 is not valid product with SKU %2". I think log details is acceptable and will help/save time during debug. @akaplya  what do you think?

Reply:
Sounds reasonable